### PR TITLE
fix(curriculum): Improved Use Recursion to Create a Countdown challenge

### PR DESCRIPTION
**Pull request by RandellDawson**
- [x] I have read [freeCodeCamp's contribution guidelines](https://github.com/freeCodeCamp/freeCodeCamp/blob/master/CONTRIBUTING.md).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `master` branch of freeCodeCamp.
- [x] None of my changes are plagiarized from another source without proper attribution.
- [x] All the files I changed are in the same world language (for example: only English changes, or only Chinese changes, etc.)
- [x] My changes do not use shortened URLs or affiliate links.

Since the new challenge [Use Recursion to Create a Countdown]() was released to production, there have been many confused users on how the challenge is supposed to work.  The confusion stems from the fact that we have a helper function behind the scenes which is technically apart of the challenge tests.  The function `countdown` function does not need to return anything, so users can really test it directly.  This is coupled with the fact that unless they understand the 1st argument is actually going to be an empty array behind the scenes.

I have taken a first shot at rewriting the challenge by sticking to the way we have written other such challenges which have functions that actually return values the users can test.  Any thoughts on how to improve this challenge further are welcomed.

See below for several forum discussions of this challenge.

* https://www.freecodecamp.org/forum/t/basic-javascript-use-recursion-to-create-a-countdown/322222

* https://www.freecodecamp.org/forum/t/use-recursion-to-create-a-countdown/323606

* https://www.freecodecamp.org/forum/t/confused-of-referenceerror-myarray-is-not-defined/323822

### CODE REVIEW

**Comment by Manish-Giri** [comment]
While I agree that this particular challenge needs some improvements, I wonder if it might be a bit tricky for beginners to conceptualize a recursive solution that uses `.concat()` - create an array with `n` in every call and merge it with the array being returned from the next recursive call. The examples used with `.push()` introduced in the earlier challenge(s) were pretty basic, so this might be quite a big leap (especially for beginners).

Maybe, we could add some explanation about how `myArray` is being set up and used behind the scenes in the current challenge, and (potentially) how campers are expected to use it? This seems to be what's confusing most campers, based on the forum posts and is also what confused me, tbh. This way, we can also be consistent with the method of using `.push()`, following from the previous challenge. 

But yes, this way there is no return value from the recursion, so we would need to consider which of the two approaches is better.

**Comment by RandellDawson** [comment]
@Manish-Giri The users will have already learned unshift, so the following solution could be used:
```js
function countdown(n) {
  if (n < 1) {
    return [];
  } else {
    const arr = countdown(n - 1);
    arr.unshift(n);
    return arr;
  }
}
```

**Comment by Manish-Giri** [comment]
@RandellDawson Sounds good. The example would help.

**Comment by RandellDawson** [comment]
I will wait until a couple other mods chime in before committing further changes.

**Comment by danshil** [comment]
I am a student who just tried to complete the project. The confusion stems from this problem having a global myArray variable that is being manipulated by this function. This is different from any of the preceding problems, and the fact that there is a global variable is not identified in the problem text. This leads to problems when you try to return algorithms.

A simple statement such as "There is a global variable called _myArray_ and you are adding values to this array..."
Or even better, just include a `const myArray = [];` above the function and let students figure out what's going on.

**Comment by RandellDawson** [comment]
> The fact that there is a global variable is not identified in the problem text. 

There is no global variable being used here.

**Comment by Ieahleen** [comment]
much less confusing like this... even if it has become a variation of the previous one, [Use Recursion to Create a Range of Numbers](https://www.freecodecamp.org/learn/javascript-algorithms-and-data-structures/basic-javascript/use-recursion-to-create-a-range-of-numbers)

seems a difficult situation! would there be an other simple algorithm that could use recursion, but is not the usual factorial calculation, or countdown? even if it is just making mandatory solving something like reversing a string with recursion. in this way there is some variation in the challenges, but they are not too ermetic for new learners

**Comment by RandellDawson** [comment]
@Ieahleen The trick is having an example which is similar enough in the description section of the challenge, so the solution only takes slight changes.  Otherwise, we are creating a challenge which would belong in one of the 3 algorithm sections (which defeats the purpose of having it in the Basic JavaScript section). 

Another "fairly" simple recursive problem would be to first show how recursion could be use to sum the elements of an array (like below):
```js
function sumArrayElements(arr) {
  if (arr.length <= 1) {
    return arr[0] || 0;
  }
  else {
    return arr.pop() + sumArrayElements(arr);
  }
}

sumArrayElements([5, 2, 3]); // 10
```
and then ask the user to create a recursive solution to multiply the elements of an array (finding the product):
```js
function multiplyArrayElements(arr) {
  if (arr.length <= 1) {
    return arr[0] || 0;
  }
  else {
    return arr.pop() * multiplyArrayElements(arr);
  }
}

multiplyArrayElements([5, 2, 3]); // 30
```
However, this seem to easy (only have to change `+` to a `*` to solve.

**Comment by ojeytonwilliams** [comment]
It seems to me that the [previous challenge](https://www.freecodecamp.org/learn/javascript-algorithms-and-data-structures/basic-javascript/use-recursion-to-create-a-range-of-numbers) is the trickier one, since it's a function of two variables.  At least I had more difficulty with it when I re-tried it.  Might be better to have that function be the final challenge.

So, what I think is that the second recursion challenge should be to create 

```js
function countup(n) {
  if (n < 1) {
    return [];
  } else {
    const countArray = countup(n - 1);
    countArray.push(n);
    return countArray ;
  }
}
```

and the final challenge should be to create

```js
function rangeOfNumbers(startNum, endNum) {
  if (endNum < startNum) {
    return [];
  } else {
    var numbers = rangeOfNumbers(startNum, endNum - 1);
    numbers.push(endNum);
    return numbers;
  }
}
```

This would make them quite similar, but I think recursion is a tricky enough concept that doing a few variations on a theme would not hurt.

**Comment by Writhyn** [comment]
Also as someone trying to solve this, my problem was that it almost sounded like my function was supposed to *create* the myArray array. There was no indication that myArray was already defined as an empty array in the system to be used for the tests. Simply putting `const myArray = [];` above the "change code below this line" would make this crystal clear with no other changes.

**Comment by RandellDawson** [comment]
@ojeytonwilliams I shifted things around, so the 2nd to last recursion challenge is the Use Recursion to Create a Countdown challenge.  It has an example in it along with the content that was originally in the now 3rd challenge (Use Recursion to Create a Range of Numbers).  Do we need an example in the 3rd one now or is the information and example in the now 2nd challenge enough for the user to be able to solve the last one without an example?  I feel any example using a range of any sort would give too much away.

**Review by raisedadead — APPROVED** [review]
Platform code LGTM.

**Review by moT01 — CHANGES_REQUESTED** [review]
Overall, this looks good - I made a number of suggestions, there's a few that are my opinion and could maybe be left alone - others that might have been preexisting. But some of them should probably be fixed. Let me know if you if you disagree.

File: curriculum/challenges/english/02-javascript-algorithms-and-data-structures/basic-javascript/use-recursion-to-create-a-countdown.md [hunk 1]
```diff
@@ -26,12 +45,12 @@ Your function must use recursion by calling itself and must not use loops of any
...
-    testString: assert.deepStrictEqual(padArray([], 10), [10, 9, 8, 7, 6, 5, 4, 3, 2, 1]);
-  - text: After calling <code>countdown(myArray, 5)</code>, myArray should contain <code>[5, 4, 3, 2, 1]</code>
-    testString: assert.deepStrictEqual(padArray([], 5), [5, 4, 3, 2, 1]);
+  - text: <code>countdown(-1)</code> should return an empty array.
+    testString: assert.isEmpty(countdown(-1));
+  - text: <code>countdown(10)</code> should return <code>[10, 9, 8, 7, 6, 5, 4, 3, 2, 1]</code>
+    testString: assert.deepStrictEqual(countdown(10), [10, 9, 8, 7, 6, 5, 4, 3, 2, 1]);
+  - text: <code>countdown(5)</code>, should return <code>[5, 4, 3, 2, 1]</code>
+    testString: assert.deepStrictEqual(countdown(5), [5, 4, 3, 2, 1]);
   - text: Your code should not rely on any kind of loops (<code>for</code> or <code>while</code> or higher order functions such as <code>forEach</code>, <code>map</code>, <code>filter</code>, or <code>reduce</code>.).
```
[the remark this answers is not in the document]
Comment (RandellDawson) [reply curriculum/challenges/english/02-javascript-algorithms-and-data-structures/basic-javascript/use-recursion-to-create-a-countdown.md]: @moT01 It seems your suggestion also removes a necessary space before the `- text` that will cause the tests to fail if removed.

[the remark this answers is not in the document]
Comment (moT01) [reply curriculum/challenges/english/02-javascript-algorithms-and-data-structures/basic-javascript/use-recursion-to-create-a-countdown.md]: I modified it.

**Comment by RandellDawson** [comment]
@moT01 Changes made.

**Review by moT01 — APPROVED** [review]
Looks like that space made travis fail.

File: curriculum/challenges/english/02-javascript-algorithms-and-data-structures/basic-javascript/use-recursion-to-create-a-countdown.md [hunk 2]
```diff
@@ -26,13 +45,13 @@ Your function must use recursion by calling itself and must not use loops of any
...
-  - text: After calling <code>countdown(myArray, 5)</code>, myArray should contain <code>[5, 4, 3, 2, 1]</code>
-    testString: assert.deepStrictEqual(padArray([], 5), [5, 4, 3, 2, 1]);
-  - text: Your code should not rely on any kind of loops (<code>for</code> or <code>while</code> or higher order functions such as <code>forEach</code>, <code>map</code>, <code>filter</code>, or <code>reduce</code>.).
+  - text: <code>countdown(-1)</code> should return an empty array.
+    testString: assert.isEmpty(countdown(-1));
+  - text: <code>countdown(10)</code> should return <code>[10, 9, 8, 7, 6, 5, 4, 3, 2, 1]</code>
+    testString: assert.deepStrictEqual(countdown(10), [10, 9, 8, 7, 6, 5, 4, 3, 2, 1]);
+  - text: <code>countdown(5)</code> should return <code>[5, 4, 3, 2, 1]</code>
+    testString: assert.deepStrictEqual(countdown(5), [5, 4, 3, 2, 1]);
+   - text: Your code should not rely on any kind of loops (<code>for</code>, <code>while</code> or higher order functions such as <code>forEach</code>, <code>map</code>, <code>filter</code>, and <code>reduce</code>).
```
Comment (moT01) [inline curriculum/challenges/english/02-javascript-algorithms-and-data-structures/basic-javascript/use-recursion-to-create-a-countdown.md]:
```suggestion
  - text: Your code should not rely on any kind of loops (<code>for</code>, <code>while</code> or higher order functions such as <code>forEach</code>, <code>map</code>, <code>filter</code>, and <code>reduce</code>).
```

**Review by moT01 — CHANGES_REQUESTED** [review]

**Review by moT01 — APPROVED** [review]
LGTM 🎉 - hey @ojeytonwilliams, you have an open PR that works on some of these files I think - do you want to take a quick look at this and make sure it's not going to create any complications.

**Comment by ojeytonwilliams** [comment]
@moT01 it's fine, I'll just review and merge this.  If there are problems with my PR, I'll sort them out.

**Review by ojeytonwilliams — APPROVED** [review]
LGTM :+1:

**Pull request merged**